cosmwasm/enclaves/execute/src/registration/onchain.rs

```diff
@@ -56,21 +120,83 @@ pub unsafe extern "C" fn ecall_authenticate_new_node(
         return NodeAuthResult::SignatureInvalid;
     }
 
-    let result = panic::catch_unwind(|| -> Result<Vec<u8>, NodeAuthResult> {
+    let mut target_public_key: [u8; 32] = [0u8; 32];
+
+    let (vec_cert, vec_quote, vec_coll) = split_combined_cert(cert, cert_len);
+
+    if vec_quote.is_empty() || vec_coll.is_empty() {
+
+        if vec_cert.is_empty() {
+            warn!("No valid attestation method provided");
+            return NodeAuthResult::InvalidCert;
+        }
+
+        trace!("EPID attestation");
+
+        //let result = panic::catch_unwind(|| {
         // verify certificate, and return the public key in the extra data of the report
-        let pk = verify_ra_cert(cert_slice, None, true)?;
+        let pk = match verify_ra_cert(cert_slice, None, true) {
+            Ok(retval) => {
+                retval
+            }
+            Err(e) => {
+                return e;
+            }
+        };
 
         // just make sure the length isn't wrong for some reason (certificate may be malformed)
         if pk.len() != PUBLIC_KEY_SIZE {
             warn!(
-                "Got public key from certificate with the wrong size: {:?}",
-                pk.len()
-            );
-            return Err(NodeAuthResult::MalformedPublicKey);
+                    "Got public key from certificate with the wrong size: {:?}",
+                    pk.len()
+                );
+            return NodeAuthResult::MalformedPublicKey;
         }
 
-        let mut target_public_key: [u8; 32] = [0u8; 32];
         target_public_key.copy_from_slice(&pk);
+        //    NodeAuthResult::Success // not yet actually
+        //});
+
+        //if result.is_err() {
+        //    // There's no real need here to test if oom happened
+        //    get_then_clear_oom_happened();
+        //    warn!("Enclave call ecall_authenticate_new_node panic!");
+        //    return NodeAuthResult::Panic;
+        //}
+
+    } else {
+
+        // DCAP
+        trace!("DCAP attestation");
+
+        let tm_s = get_current_block_time_s();
+        trace!("Current block time: {}", tm_s);
+
+        // test self
+        let report_body = match verify_quote_ecdsa(&vec_quote, &vec_coll, tm_s) {
+            Ok(r) => {
+                trace!("Remote quote verified ok");
+                if r.1 != sgx_ql_qv_result_t::SGX_QL_QV_RESULT_OK {
+                    trace!("WARNING: {}", r.1);
+                }
+                r.0
+            }
+            Err(e) => {
+                trace!("Remote quote verification failed: {}", e);
+                return NodeAuthResult::InvalidCert;
+            }
+        };
```

Comment:
Question about this - if dcap isn't properly installed will this fail? Do we have any way to guarantee that if the node starts there aren't any driver dependencies here that will cause an error? I'm asking because the results of this function have to be the same for each and every node. Anyone that doesn't conform with the consensus will crash (or even halt the chain if there are too many)

Reply:
Theoretically this should always succeed (but need to check). This method does't demand a driver.